File: base/sort/python/MergeSort.py
# 时间复杂度O(nlog2n),最坏情况O(nlog2n)，最好情况O(nlog2n),空间复杂度O(n), 稳定
def merge_sort(unsort_nums):
    sort_nums = _merge_sort(unsort_nums)
    for idx, num in enumerate(unsort_nums):
        unsort_nums[idx] = sort_nums[idx]


def _merge_sort(unsort_nums):
    if len(unsort_nums) < 2:
        return unsort_nums
    mid = len(unsort_nums) // 2
    left_nums = unsort_nums[:mid]
    right_nums = unsort_nums[mid:]
    return _merge(_merge_sort(left_nums), _merge_sort(right_nums))


def _merge(left, right):
    result = []
    while left and right:
        if left[0] <= right[0]:
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))

    if left:
        result.extend(left)
    if right:
        result.extend(right)
    return result

File: base/sort/java/MergeSort.java
package sort.java;


//归并排序
public class MergeSort {

    public static void mergeSort(int[] array) {
        if (array == null || array.length <= 1) {
            return;
        }
        int[] sortArray = merge(array);
        System.arraycopy(sortArray, 0, array, 0, sortArray.length);
    }

    private static int[] merge(int[] array) {
        if (array.length < 2) return array;
        int mid = array.length / 2;
        int[] left = new int[mid], right = new int[array.length - mid];
        System.arraycopy(array, 0, left, 0, mid);
        System.arraycopy(array, mid, right, 0, array.length - mid);
        return _merge(merge(left), merge(right));
    }

    private static int[] _merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i = 0, j = 0;
        while (i < left.length && j < right.length) {
            if (left[i] > right[j]) result[i + j] = right[j++];
            else result[i + j] = left[i++];
        }
        if (i < left.length) System.arraycopy(left, i, result, i + j, left.length - i);
        if (j < right.length) System.arraycopy(right, j, result, i + j, right.length - j);
        return result;
    }
}
